fix exception handling in bilby loader (no ticket)

geishm-ansto opened the pull request:
This is an exception handling fix to a hardware panel failure that led to the corruption of decompressed 
tof data for the good events.
  
There are no additional tests cases as this is a fix in response to a hardware failure.
The existing tests cases should be sufficient to check performance with valid datasets is unchanged.


No release info needed.

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

Commit:
fix exception handling in bilby loader (no ticket)

Pasarus commented on Framework/DataHandling/src/LoadBBY.cpp:
```diff
@@ -783,7 +783,9 @@ void LoadBBY::loadEvents(API::Progress &prog, const char *progMsg, ANSTO::Tar::F
         eventProcessor.newFrame();
       } else if ((x >= HISTO_BINS_X) || (y >= HISTO_BINS_Y)) {
-        // ignore
+        // cannot ignore the dt contrbition even if the event
```
```suggestion
        // cannot ignore the dt contribution even if the event
```
Is this a typo?

Pasarus reviewed (changes requested):
This does need release notes, as it changes the output of the program and users need to be informed of these changes.

Also this code needs to be both tested manually and automatically if possible, could you provide a dataset that causes this bug without this code? Then proving that this actually works.
In terms of a unit test could you fake the failure by editing some data or providing a fake instrument with the hardware failure baked in? That way this won't regress in the future.

geishm-ansto commented:
This correction is a fix to a recent hardware failure that invalidated the decompression algorithm in the loader. Any files existing customers have do not have this problem because the data would have been unusable and rejected. Reprocessing their existing files does not change the output as the regression tests demonstrate. 

The purpose of this code change was to recover data from an experiment that cannot be repeated now. I do not understand the logic of adding automated system and unit test cases and additional test datasets to demonstrate that the software can handle such a hardware failure. I am happy to add warnings to the log if that code is reached to indicate that invalid events were detected and a unit test to demonstrate that the invalid events are logged. If this is acceptable a release note can be added to indicate the logging of invalid events.

Pasarus commented:
We use Unit Tests to test against feature regression, if you want this to work into the future we need to add a unit test to reduce the likelihood that this occurs again for more users should a developer change this code in the future, is there a way for you to simulate the hardware failure? Which part of the code falls over? Why can't we just fake it, for the purpose of the test? This is about good software practices and developing good software, not just about fixing one case, this one time.

Also this needs release notes, as it's still a change to Mantid that a user can see.

Pasarus commented:
> This correction is a fix to a recent hardware failure that invalidated the decompression algorithm in the loader. Any files existing customers have do not have this problem because the data would have been unusable and rejected. Reprocessing their existing files does not change the output as the regression tests demonstrate.
> 
> The purpose of this code change was to recover data from an experiment that cannot be repeated now. I do not understand the logic of adding automated system and unit test cases and additional test datasets to demonstrate that the software can handle such a hardware failure. I am happy to add warnings to the log if that code is reached to indicate that invalid events were detected and a unit test to demonstrate that the invalid events are logged. If this is acceptable a release note can be added to indicate the logging of invalid events.

I just want to make it clear that I am happy as long as there is the test that you suggested. I don't think that came across in the original comment. The release notes would also be much appreciated. Would it be possible though to add the data that is "lost" or corrupted as a unit test data? (Obviously if there is no embargo or required secrecy) I'd be happy to do that test by hand one time and not in a unit test or disclosing the data outside of ISIS before approving this PR if you can email it to me.

Commit:
added unit test to catch invalid value (no ticket)

geishm-ansto commented:
I added a unit test case and updated the release note to advise that the Bilby loader logs the occurence of invalid events as a warning. I didn't add any new files as the test could be demonstrated by copying the existing Bilby test file and corrupting a single event in the file.

Pasarus reviewed (approved):
I am happy with the changes made, the unit test is about as good as we can get, given the circumstances, and informing users that their data is invalid is valuable.

Approved :+1:
